index.js

```diff
@@ -3073,38 +3087,78 @@
   //.
   //. > S.last ([])
   //. Nothing
+  //.
+  //. > S.last (Cons ('baz') (Cons ('foo') (Nil)))
+  //. S.Just ('foo')
+  //.
+  //. > S.last (Nil)
+  //. S.Nothing
   //. ```
-  function last(xs) {
-    return xs.length > 0 ? Just (xs[xs.length - 1]) : Nothing;
-  }
+
+  function last(foldable) {
+    return Z.reduce (
+      function(acc, a) {
+        return Just(a);
+      },
+      Nothing,
+      foldable
+     );
+    }
+
   _.last = {
-    consts: {},
-    types: [$.Array (a), $Maybe (a)],
+    consts: {f: [Z.Foldable]},
+    types: [f (a), $Maybe (a)],
     impl: last
   };
 
-  //# tail :: Array a -> Maybe (Array a)
+  //# tail :: (Foldable f, Applicative f) => f a -> Maybe (f a)
```

Comment:
Could you justify the Applicative constraint?

Reply:
In order to build a new foldable with the right order I need to use 'append' which requires that f be an Applicative and Semigroup. I'm now a bit confused about why it is working without both being required.

Reply:
> In order to build a new foldable with the right order I need to use 'append' which requires that f be an Applicative and Semigroup.

Good point! It looks as though you're also using [`Z.empty`][1], so there may be other constraints to consider.

> I'm now a bit confused about why it is working without both being required.

Type classes need not be specified in order for a function to work. In fact, the only reason to specify a function's types and type-class constraints is to ensure that *invalid* applications of the function result in errors with descriptive messages. For example:

```javascript
S.concat (123) (456);
// ! TypeError: Type-class constraint violation
//
//   concat :: Semigroup a => a -> a -> a
//             ^^^^^^^^^^^    ^
//                            1
//
//   1)  123 :: Number, FiniteNumber, NonZeroFiniteNumber, Integer, NonNegativeInteger, ValidNumber
//
//   ‘concat’ requires ‘a’ to satisfy the Semigroup type-class constraint; the value at position 1 does not.
//
//   See https://github.com/sanctuary-js/sanctuary-type-classes/tree/v10.0.0#Semigroup for information about the sanctuary-type-classes/Semigroup type class.
```

If we replace `{a: [Z.Semigroup]}` with `{}` and try again, the error is less helpful:

```javascript
S.concat (123) (456);
// ! TypeError: Semigroup.methods.concat(...) is not a function
```


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/tree/v10.0.0#empty

Reply:
Verbose messages are much more helpful.

Reply:
> It looks as though you're also using [`Z.empty`][1], so there may be other constraints to consider.

Have you given this some thought?


[1]: https://github.com/sanctuary-js/sanctuary-type-classes/tree/v10.0.0#empty

Reply:
Now I have :)

Reply:
🆗